alica_engine/src/engine/BasicPlan.cpp

```diff
@@ -13,35 +13,74 @@ BasicPlan::BasicPlan()
         , _configuration(nullptr)
         , _msInterval(AlicaTime::milliseconds(DEFAULT_MS_INTERVAL))
         , _activeRunJobId(-1)
+        , _isMasterPlan(false)
         , _signalContext(nullptr)
         , _execContext(nullptr)
         , _signalState(1)
         , _execState(1)
-        , _flags(static_cast<uint8_t>(Flags::TRACING_ENABLED))
+        , _tracingType(TracingType::DEFAULT)
+        , _customTraceContextGetter()
+        , _trace()
+        , _runTraced(false)
+        , _initExecuted(false)
 {
 }
 
+void BasicPlan::startTrace()
+{
+    if (!_ae->getTraceFactory()) {
+        return;
+    }
+
+    switch (_tracingType) {
+    case TracingType::DEFAULT: {
+        auto parent = _execContext.load()->getParent();
+        for (; parent && (!parent->getBasicPlan() || !parent->getBasicPlan()->getTraceContext().has_value()); parent = parent->getParent());
+        _trace = _ae->getTraceFactory()->create(_name, parent ? parent->getBasicPlan()->getTraceContext() : std::nullopt);
+        break;
+    }
+    case TracingType::SKIP: {
+        break;
+    }
+    case TracingType::ROOT: {
+        _trace = _ae->getTraceFactory()->create(_name);
+        break;
+    }
+    case TracingType::CUSTOM: {
+        if (!_customTraceContextGetter) {
+            ALICA_ERROR_MSG("[BasicPlan] Custom tracing type specified, but no getter for the trace context is provided. Switching to default tracing type instead");
+            _tracingType = TracingType::DEFAULT;
+            startTrace();
+            break;
+        }
+        _trace = _ae->getTraceFactory()->create(_name, _customTraceContextGetter(this));
+        break;
+    }
+    }
+
+    if (_isMasterPlan && _trace) {
+        // Immediately send out the trace for the master plan, otherwise the traces will not be available till the application end
+        _trace->finish();
```

Comment:
can't we just destruct the trace here? what's the use of keeping object alive.

Reply:
it is to get the trace context in the child plans. If we destroy it here, then need to store the trace context separately

Reply:
Sure but what's the guarantee that context() will return valid context even after a finish has been called, since the implementation will be inside different package.

Reply:
as per open tracing the context remains valid even after the trace is finished. So I followed the same semantics for the IAlicaTrace interface.